### PR TITLE
NE-2651 - feat(jira): Add generate-bug-reproducer command with --apply for live cluster execution

### PR DESCRIPTION
**Pull request by davidesalerno**
## Summary

- Adds a new `/jira:generate-bug-reproducer` command that analyzes a bug's JIRA description and fix PR code changes to produce a structured reproducer report with pre-fix reproduction steps, post-fix verification, and a confidence assessment
- Refactors `/jira:generate-test-plan` to use a shared skill (`plugins/jira/skills/generate-test-plan/SKILL.md`), eliminating 164 lines of inline implementation and enabling both commands to share the same logic
- Introduces `--apply` flag on both commands to execute the generated steps against a live OpenShift cluster

## New Command: `/jira:generate-bug-reproducer`

An alias for `/jira:generate-test-plan --reproducer`. Analyzes a JIRA bug and its fix PRs to produce a reproducer report when JIRA lacks complete reproduction steps.

**Usage:**
```bash
# Auto-discover PRs from JIRA
/jira:generate-bug-reproducer OCPBUGS-12345

# Analyze specific fix PRs
/jira:generate-bug-reproducer OCPBUGS-12345 https://github.com/openshift/hypershift/pull/6888
```

**Reproducer report includes:**
1. Bug Summary
2. Root Cause Analysis (inferred from PR diff)
3. Prerequisites / Environment
4. Steps to Reproduce (Pre-Fix)
5. Observed vs Expected Behavior
6. Fix Verification Steps (Post-Fix)
7. Automation Opportunities
8. Confidence & Assumptions (high/medium/low)
9. Open Questions
10. PR Summary

## New Feature: `--apply` Flag

Executes the generated reproducer or test plan steps against a live OpenShift cluster using the `oc` CLI. Available on both `/jira:generate-bug-reproducer` and `/jira:generate-test-plan`.

**Usage:**
```bash
# Generate reproducer and apply to live cluster
/jira:generate-bug-reproducer OCPBUGS-12345 --apply

# Generate test plan and apply scenarios to cluster
/jira:generate-test-plan CNTRLPLANE-205 --apply
```

**How it works:**

1. **Report reuse** — if a report already exists for the JIRA key, offers to apply it directly without regenerating
2. **Prerequisite checks** — verifies `oc` CLI is installed, cluster is reachable (`oc whoami`), and displays cluster context for user confirmation
3. **Command classification** — classifies each step as read-only (`oc get`, `oc describe`, `oc logs`) or write (`oc apply`, `oc create`, `oc delete`, `oc patch`, etc.)
4. **Hybrid confirmation model**:
   - One-time bulk approval for all read-only commands
   - Individual confirmation before each write command (yes/skip/abort)
5. **Output comparison** — captures command output and compares with expected behavior from the report
6. **Results summary** — displays per-step results (EXECUTED/SKIPPED/NOT REACHED), bug reproduction assessment, and lists any resources created but not cleaned up
7. **Execution log** — saves results to `.work/jira/generate-test-plan/{ISSUE_KEY}/apply-log-{timestamp}.md`

## Shared Skill Refactoring

Extracted the implementation from `generate-test-plan.md` into a shared skill at `plugins/jira/skills/generate-test-plan/SKILL.md`. Improvements over the old inline implementation:

- **Restored test scenario categories** — explicit guidance for happy path, negative, edge cases, integration, error handling, performance, and platform-specific test variations (lost during initial extraction)
- **Restored multi-PR integration guidance** — explicit instruction to create integrated test scenarios when multiple PRs fix one issue
- **Decoupled mode detection** — skill no longer checks the invoking command name; mode is set by the command or `--reproducer` flag

## Files Changed

| File | Change |
|------|--------|
| `plugins/jira/skills/generate-test-plan/SKILL.md` | **New** — shared skill with 8 implementation steps |
| `plugins/jira/commands/generate-bug-reproducer.md` | **New** — reproducer command definition |
| `plugins/jira/commands/generate-test-plan.md` | Refactored to delegate to shared skill |
| `plugins/jira/README.md` | Added documentation for both commands and `--apply` |
| `AGENTS.md` | Updated Available Plugins table |
| `README.md` | Added Bug Reproducers entry |
| `plugins/jira/commands/create.md` | Added See Also cross-reference |

## Test plan

- [ ] Verify `make lint` passes (confirmed: 0 errors, 0 warnings)
- [ ] Verify `/jira:generate-bug-reproducer OCPBUGS-XXXXX` generates a reproducer report
- [ ] Verify `/jira:generate-test-plan CNTRLPLANE-XXX` still generates a test plan
- [ ] Verify `/jira:generate-bug-reproducer OCPBUGS-XXXXX --apply` executes reproducer steps against a connected cluster
- [ ] Verify `--apply` with an existing report offers reuse option
- [ ] Verify `--apply` aborts gracefully when `oc` is not installed or cluster is unreachable
- [ ] Verify write commands require individual confirmation while read-only commands use bulk approval

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a new Jira “bug reproducer” command to generate structured reproduction steps from a JIRA issue and related fix PR, with optional live-cluster execution.
  * Expanded Jira test-plan generation with reproducer mode and support for executing generated steps against a live OpenShift cluster.
* **Documentation**
  * Updated command and skill documentation with new arguments, examples (including live `--apply` usage), and safer execution guidance.
  * Bumped the Jira plugin version to 0.9.0 in marketplace metadata.

### CODE REVIEW

**Comment by bentito** [comment]
@davidesalerno would it make sense to retitle this PR to start with "[NE-2651](https://redhat.atlassian.net/browse/NE-2651): " so that it gets tied to that story?

File: plugins/jira/commands/generate-bug-reproducer.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+## Synopsis
+```
+/jira:generate-bug-reproducer [JIRA issue key] [GitHub PR URLs] [--apply]
+```
+
+## Description
+
+The `jira:generate-bug-reproducer` command analyzes a bug's JIRA description and fix PR code changes to produce a structured reproducer report. The report includes pre-fix reproduction steps (inferred from the PR diff when JIRA lacks them), post-fix verification steps, and a confidence assessment.
+
+This command is an alias for `/jira:generate-test-plan --reproducer`. Both invoke the same **Generate Test Plan** skill in reproducer mode.
```
Comment (candita) [inline plugins/jira/commands/generate-bug-reproducer.md:18]: I'm not sure that reusing the generate-test-plan skill is all that we need here.  It is a good start, but we probably need to go a step further than that.  We don't want to be testing the same things as the developer tests.  Perhaps asking Claude to review the generated test plan and enhance it with the actual/expected details in the bug report. Or list all of the conditions that were present at the time the bug happened (as seen in the must-gather) and act on a way to make these conditions reproduceable.

File: README.md
Comment (candita) [inline README.md:1]: Can you explain why the README changed so much?

File: plugins/jira/commands/generate-test-plan.md [hunk 2]
```diff
@@ -1,186 +1,89 @@
...
+Invoke the **Generate Test Plan** skill (`plugins/jira/skills/generate-test-plan/SKILL.md`) with:
 
-5. **Build/Deploy Section Exclusion**:
-   - Does NOT include build or deployment steps
-   - Assumes environment is already set up
-   - Focuses purely on testing procedures
+- **Mode**: `test-plan` (default) or `reproducer` when `--reproducer` is present
+- **JIRA issue key**: `$1`
+- **PR URLs**: remaining arguments except `--reproducer` and `--apply`
+- **Apply mode**: `true` when `--apply` is present, `false` otherwise
```
Comment (candita) [inline plugins/jira/commands/generate-test-plan.md]: I'm curious - where is "apply" implemented?

File: plugins/jira/commands/generate-bug-reproducer.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+jira:generate-bug-reproducer
+
+## Synopsis
+```
+/jira:generate-bug-reproducer [JIRA issue key] [GitHub PR URLs] [--apply]
+```
+
+## Description
+
+The `jira:generate-bug-reproducer` command analyzes a bug's JIRA description and fix PR code changes to produce a structured reproducer report. The report includes pre-fix reproduction steps (inferred from the PR diff when JIRA lacks them), post-fix verification steps, and a confidence assessment.
```
Comment (rhamini3) [inline plugins/jira/commands/generate-bug-reproducer.md]: the expectation usually is that the pre-fix reproduction steps is basically identical to the post-fix verification just with a different expected result.

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 1 above)
Comment (rhamini3) [reply plugins/jira/commands/generate-bug-reproducer.md:18]: > enhance it with the actual/expected details in the bug report.

referring to [comment](https://github.com/openshift-eng/ai-helpers/pull/596/changes#r3516311737) above, this will be the best way to get a good reproducer to use pre and post fix imo

File: plugins/jira/commands/generate-bug-reproducer.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+## Implementation
+
+Invoke the **Generate Test Plan** skill (`plugins/jira/skills/generate-test-plan/SKILL.md`) with:
+
+- **Mode**: `reproducer`
+- **JIRA issue key**: `$1`
+- **PR URLs**: `$2`, `$3`, ... (optional; auto-discovered from JIRA if omitted)
+- **Apply mode**: `true` when `--apply` is present, `false` otherwise
+
+The skill handles JIRA fetching, PR discovery, diff analysis, report generation, and output presentation. When `--apply` is set, it also executes the reproducer steps against the connected cluster (see Step 8 in the skill).
```
Comment (rhamini3) [inline plugins/jira/commands/generate-bug-reproducer.md:33]: it also analyzes the JIRA description to pull the reproducer steps from there right? might be good to mention that here then.

File: plugins/jira/commands/generate-bug-reproducer.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   /jira:generate-bug-reproducer OCPBUGS-12345 https://github.com/openshift/hypershift/pull/6888 --apply
+   ```
+
+## Arguments
+
+- **$1**: JIRA issue key (required) — e.g., `OCPBUGS-12345`, `CNTRLPLANE-205`
+- **$2, $3, ..., $N**: Optional GitHub PR URLs
+  - If provided: only these PRs are analyzed
+  - If omitted: PRs linked to the JIRA are discovered automatically
+- **--apply**: Execute the reproducer steps against a live OpenShift cluster after generating (or reusing) the report. Requires `oc` CLI and active cluster connectivity. Uses hybrid confirmation: bulk approval for read-only commands, individual approval for write commands.
```
Comment (rhamini3) [inline plugins/jira/commands/generate-bug-reproducer.md:73]: will this only work on a CLI with a cluster already logged into?

File: plugins/jira/commands/generate-test-plan.md [hunk 6]
```diff
@@ -1,186 +1,89 @@
...
+   ```
+
+5. **Generate reproducer and apply to cluster**:
+   ```
+   /jira:generate-test-plan OCPBUGS-12345 --reproducer --apply
+   ```
+
+6. **Generate test plan and apply scenarios to cluster**:
+   ```
+   /jira:generate-test-plan CNTRLPLANE-205 --apply
```
Comment (rhamini3) [inline plugins/jira/commands/generate-test-plan.md]: is there a scenario to provide a specific PR with `--reproducer` and `--apply` ?

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 5 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-bug-reproducer.md:73]: Yes, having an `oc cli` connected to a cluster it is a prerequisite

File: plugins/jira/commands/generate-test-plan.md (hunk 2 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-test-plan.md]: Good question — `--apply` isn’t implemented in this command file itself. This file only documents the flag and passes `Apply mode: true|false` into the shared `Generate Test Plan` skill.

The actual `--apply` flow lives in `plugins/jira/skills/generate-test-plan/SKILL.md`, specifically Step 8 (“Apply to Cluster”), where it covers the prerequisite checks, command classification, confirmations, execution, and result reporting.

I can make that clearer here by saying this command delegates `--apply` handling to the shared skill and pointing to Step 8 explicitly if you think that i's needed.

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 3 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-bug-reproducer.md]: yes - but there are some cases where pre-fix reproduction steps are less (due to the bug) than the post-fix verification steps and I specified this detail to cover these cases.

Do you think it is better to remove this distinction?

File: README.md
Comment (davidesalerno) [reply README.md:1]: Good catch — the README changed because during the rebase conflict resolution the wrong side was kept (my old branch version instead of upstream's current one). I've just pushed a fix that resets it to match upstream main. The PR shouldn't change README.md at all.

File: plugins/jira/commands/generate-test-plan.md (hunk 6 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-test-plan.md]: Yes - for example when a JIRA bug has several linked PRs (even if it isn't so common) but you want to generate and execute a reproducer focused on a single specific fix PR.

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 1 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-bug-reproducer.md:18]: @candita I decided to create a new command that is reusing what we already have got into the existing `generate-test-plan` in order to not have 2 things that are doing the same to maintain but creating a new command (`generate-bug-reproducer`) that it is an alias that will let us to use only what it is needed if we want only a repoducer and not a full test plan.

> Perhaps asking Claude to review the generated test plan

 This could be a good improvement to have better results.

> and enhance it with the actual/expected details in the bug report. 

Aren't the pre-fix and the post-fix steps the same as the actual/expected?

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 3 above)
Comment (rhamini3) [reply plugins/jira/commands/generate-bug-reproducer.md]: for the specific cases, where it is different, I think you can leave it, but I think it is good to mention also that for most cases the verification is identical, if you think its worthwhile

File: plugins/jira/commands/generate-test-plan.md (hunk 6 above)
Comment (rhamini3) [reply plugins/jira/commands/generate-test-plan.md]: do you think we should add that as an example? I feel like when it comes to using this command for PR verification, usually, we verify PRs one by one, especially for bug fixes (not so much for feature verifcation)

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 5 above)
Comment (rhamini3) [reply plugins/jira/commands/generate-bug-reproducer.md:73]: do you want to mention this somewhere? maybe the pre-requisites section?

File: plugins/jira/skills/generate-test-plan/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+
+Use **test plan mode** when QE or a developer needs step-by-step guidance to validate a fix after it is applied, including regression coverage.
+
+Use **reproducer mode** when the JIRA bug lacks complete reproduction steps and you need to infer how to trigger the bug from the fix PR diff and JIRA narrative, then verify the fix resolves it.
+
+## Prerequisites
+
+- JIRA credentials (`JIRA_USERNAME`, `JIRA_API_TOKEN`) or Jira MCP server configured
+- `gh` CLI authenticated for GitHub PR access
+- `jq` available for JSON parsing
```
Comment (rhamini3) [inline plugins/jira/skills/generate-test-plan/SKILL.md:25]: here would be a good optional step to have a cluster CLI connectivity to use `--apply`

File: plugins/jira/skills/generate-test-plan/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+1. Extract the JIRA issue key from `$1`
+2. Collect optional PR URLs from remaining arguments
+3. Detect `--reproducer` flag in any argument position
+4. Set mode:
+   - `reproducer` if `--reproducer` is present or the mode is explicitly set to `reproducer` by the invoking command
+   - `test-plan` otherwise
+5. Detect `--apply` flag in any argument position
+6. Set apply mode:
+   - `true` if `--apply` is present
+   - `false` otherwise
```
[the remark this answers is not in the document]
Comment (davidesalerno) [reply plugins/jira/skills/generate-test-plan/SKILL.md:52]: @coderabbitai The current code at lines 40–49 already implements exactly what the finding describes:

  - Line 41: Extracts JIRA key from $1
  - Line 42: "Parse each remaining argument exactly once, in order"
  - Lines 43–46: Recognizes --reproducer, --apply, and valid PR URLs before anything else; rejects unknown arguments with a usage error and explicitly says "never forward an unrecognized flag to gh pr view/gh pr diff"
  - Lines 47–49: Sets mode to reproducer if the flag was seen or the invoking command set it, test-plan otherwise
  - Lines 50–52: Sets apply mode based on the --apply flag

File: plugins/jira/skills/generate-test-plan/SKILL.md (hunk 7 above)
Comment (davidesalerno) [reply plugins/jira/skills/generate-test-plan/SKILL.md:25]: I addressed this (see at line 26)

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 4 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-bug-reproducer.md:39]: I',m adding  a `Prerequisites` section between `Synopsis` and `Description` in                            `plugins/jira/commands/generate-bug-reproducer.md` with exactly that content

File: plugins/jira/commands/generate-bug-reproducer.md (hunk 3 above)
Comment (davidesalerno) [reply plugins/jira/commands/generate-bug-reproducer.md]: I think that it is mentioned at line 22:                                                                                                                                                  
                                                                                                                                                                                                                                    
  ▎ "In most cases the verification steps mirror the reproduction steps exactly, just with a different expected 

What do you think?
